### PR TITLE
Refactor TypeChecker to assign different IDs to different errors

### PR DESCRIPTION
**Pull request by a3d4**
There are still cases where different errors in `TypeChecker` are reported with the same ID. This PR addresses them.

Note that this is not a black-and-white picture. Sometimes it is not really clear if we have two different errors or just two flavors of of the same error.

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1994,19 +1994,16 @@ void TypeChecker::typeCheckFunctionGeneralChecks(
 		bool const isStructConstructorCall =
 			_functionCall.annotation().kind == FunctionCallKind::StructConstructorCall;
 
-		string msg;
-
-		if (isVariadic)
-			msg +=
+		auto [errorId, description] = [&]() -> tuple<ErrorId, string> {
```
Comment (mijovic) [inline libsolidity/analysis/TypeChecker.cpp:1997]: Do you need lambda function here? Maybe you can do it without lambda

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -2549,61 +2571,66 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+			string errorMsg = "Member \"" + memberName + "\" not found or not visible "
+				"after argument-dependent lookup in " + exprType->toString() + ".";
 
-			if (memberName == "value")
+			if (auto const& funType = dynamic_cast<FunctionType const*>(exprType))
 			{
-				if (funType->kind() == FunctionType::Kind::Creation)
-					errorMsg = "Constructor for " + t.front()->toString() + " must be payable for member \"value\" to be available.";
+				auto const& t = funType->returnParameterTypes();
```
Comment (mijovic) [inline libsolidity/analysis/TypeChecker.cpp]: Consider avoiding auto in this case

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -2549,61 +2571,66 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+
+				if (memberName == "value")
+				{
+					if (funType->kind() == FunctionType::Kind::Creation)
+						return { 8827_error, "Constructor for " + t.front()->toString() + " must be payable for member \"value\" to be available." };
```
Comment (mijovic) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
						return { 
							8827_error, 
							"Constructor for " + t.front()->toString() + " must be payable for member \"value\" to be available."
						};
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -2549,61 +2571,66 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+			else if (exprType->category() == Type::Category::Contract)
 			{
-				solAssert(
-					addressType->stateMutability() != StateMutability::Payable,
-					"Expected address not-payable as members were not found"
-				);
+				for (auto const& addressMember: TypeProvider::payableAddress()->nativeMembers(nullptr))
+					if (addressMember.name == memberName)
+					{
+						Identifier const* var = dynamic_cast<Identifier const*>(&_memberAccess.expression());
```
Comment (mijovic) [inline libsolidity/analysis/TypeChecker.cpp]: Good example for usage of `auto`
```suggestion
						auto const* var = dynamic_cast<Identifier const*>(&_memberAccess.expression());
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 5]
```diff
@@ -2549,61 +2571,66 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+						string varName = var ? var->name() : "...";
+						errorMsg += " Use \"address(" + varName + ")." + memberName + "\" to access this address member.";
+						return { 5256_error, errorMsg };
+					}
+			}
+			else if (auto addressType = dynamic_cast<AddressType const*>(exprType))
```
Comment (mijovic) [inline libsolidity/analysis/TypeChecker.cpp]: To be more consistent with the rest of PR where you extract `const` and `*/&` from `auto`
```suggestion
			else if (auto const* addressType = dynamic_cast<AddressType const*>(exprType))
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 6]
```diff
@@ -2549,61 +2571,66 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
-		{
-			// Trigger error when using send or transfer with a non-payable fallback function.
-			if (memberName == "send" || memberName == "transfer")
+			else if (exprType->category() == Type::Category::Contract)
 			{
-				solAssert(
-					addressType->stateMutability() != StateMutability::Payable,
-					"Expected address not-payable as members were not found"
-				);
+				for (auto const& addressMember: TypeProvider::payableAddress()->nativeMembers(nullptr))
```
Comment (mijovic) [inline libsolidity/analysis/TypeChecker.cpp]: Personally I would not use auto here, as it is hard to imagine what is particular type...

File: libsolidity/analysis/TypeChecker.cpp [hunk 7]
```diff
@@ -2549,61 +2571,66 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+						return { 8477_error, "Member \"value\" is not allowed in delegated calls due to \"msg.value\" persisting." };
+					else
+						return { 8820_error, "Member \"value\" is only available for payable functions." };
+				}
 				else if (
-					funType->kind() == FunctionType::Kind::DelegateCall ||
-					funType->kind() == FunctionType::Kind::BareDelegateCall
+					t.size() == 1 &&
+					(t.front()->category() == Type::Category::Struct ||
+					t.front()->category() == Type::Category::Contract)
```
Comment (mijovic) [inline libsolidity/analysis/TypeChecker.cpp]: Not entirely sure what is correct for our coding style. One you wrote or:
```suggestion
					t.size() == 1 && (
						t.front()->category() == Type::Category::Struct ||
						t.front()->category() == Type::Category::Contract
					)
```
or
```suggestion
					t.size() == 1 &&
					(
						t.front()->category() == Type::Category::Struct ||
						t.front()->category() == Type::Category::Contract
					)
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 8]
```diff
@@ -2549,61 +2571,66 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
 
-		if (auto const& funType = dynamic_cast<FunctionType const*>(exprType))
-		{
-			auto const& t = funType->returnParameterTypes();
+		auto [errorId, description] = [&]() -> tuple<ErrorId, string> {
+			string errorMsg = "Member \"" + memberName + "\" not found or not visible "
+				"after argument-dependent lookup in " + exprType->toString() + ".";
 
-			if (memberName == "value")
+			if (auto const& funType = dynamic_cast<FunctionType const*>(exprType))
```
Comment (mijovic) [inline libsolidity/analysis/TypeChecker.cpp]: I know it is old code, but should be fixed to:
```suggestion
			if (auto const* funType = dynamic_cast<FunctionType const*>(exprType))
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 9]
```diff
@@ -2016,49 +2013,64 @@ void TypeChecker::typeCheckFunctionGeneralChecks(
...
-			TypePointer const expressionType = type(_functionCall.expression());
-			TypeType const& t = dynamic_cast<TypeType const&>(*expressionType);
-			auto const& structType = dynamic_cast<StructType const&>(*t.actualType());
-			set<string> membersRemovedForStructConstructor = structType.membersMissingInMemory();
+			// Extend error message in case we try to construct a struct with mapping member.
+			if (isStructConstructorCall)
+			{
+				/// For error message: Struct members that were removed during conversion to memory.
+				TypePointer const expressionType = type(_functionCall.expression());
+				TypeType const& t = dynamic_cast<TypeType const&>(*expressionType);
```
Comment (mijovic) [inline libsolidity/analysis/TypeChecker.cpp]: Old code, but is good example of usage for auto:
```suggestion
				auto const& t = dynamic_cast<TypeType const&>(*expressionType);
```

**Review by mijovic — COMMENTED** [review]
Looks good in general. Few style issues and maybe removing lambdas as they are just called inline with definition.

I assume that you are sure all the error codes are unique.

File: libsolidity/analysis/TypeChecker.cpp (hunk 7 above)
Comment (a3d4) [reply libsolidity/analysis/TypeChecker.cpp]: I just kept old formatting. I believe it should be (according to the code style):

```
t.size() == 1 && (
    t.front()->category() == Type::Category::Struct ||
    t.front()->category() == Type::Category::Contract
)
```

**Review by mijovic — COMMENTED** [review]
I think it makes sense to have lambdas after doing another review round. Only open question is if it can be better with member functions. I think @chriseth will probably have some idea.

**Comment by chriseth** [comment]
Can you squash?

**Review by chriseth — APPROVED** [review]

**Comment by a3d4** [comment]
> I think the seed for the random number generator was not set correctly.

It was intentional, having reproducible random numbers is often helpful.

However I wanted to change it for the production system just the way you did. Thanks a lot!

**Pull request merged**